ball_detect/main.c

```diff
@@ -12,115 +12,134 @@
+        | (1 << DIP_SWITCH_4)
+        | (1 << DIP_SWITCH_5)
+        | (1 << DIP_SWITCH_6));
 
-	//DETECT ADRESS
-	if((PINC & (1 << DIP_SWITCH_1)) == 0) {
-		my_address = 1;
-	} else if((PINC & (1 << DIP_SWITCH_2)) == 0) {
-		my_address = 2;
-	} else if((PINC & (1 << DIP_SWITCH_3)) == 0) {
-		my_address = 3;
-	} else if((PINC & (1 << DIP_SWITCH_4)) == 0) {
-		my_address = 4;
-	}
-	
-	//TURN ON INTERRUPTS
-	sei();
+    //INTERRUPTS INIT
+    DDRB = 0;
+    PORTB |= ((1 << BALL_SENSOR_1)
+        | (1 << BALL_SENSOR_2)
+        | (1 << BALL_SENSOR_3)
+        | (1 << ULTRASONIC_SENSOR));
+    PCICR |= 1;
+    PCMSK0 |= 0xF;
+
+    //TIMER SETUP
+    TCCR1A = 0;
+    TCCR1B = 2;
+    TCNT1 = 0;
+
+    //DETECT ADRESS
+    if((PINC & (1 << DIP_SWITCH_1)) == 0) {
+        my_address = 1;
+    } else if((PINC & (1 << DIP_SWITCH_2)) == 0) {
+        my_address = 2;
+    } else if((PINC & (1 << DIP_SWITCH_3)) == 0) {
+        my_address = 3;
+    } else if((PINC & (1 << DIP_SWITCH_4)) == 0) {
+        my_address = 4;
+    }
+
+    //TURN ON INTERRUPTS
+    sei();
 }
 
 void sendc(uint8_t ch) {
-  waitForTX();
-  UDR0 = ch;
+    waitForTX();
+    UDR0 = ch;
 }
 
-ISR(USART_RX_vect) {	
-	PORTD ^= 1 << 3;
-	uint32_t modulo, position;
-	uint8_t read_char = UDR0, i, j;
-	if(read_char == my_address) {
-		for(j = 2; j >= 0; j--) {	
-			modulo = vision_result[j];
-			position = 1000000;
-			for(i = 0; i < 7; i++) {
-				sendc((modulo / position) % 10);
-				modulo %= position;
-				position /= 10;
-			}
-		}
-	} 		
+ISR(USART_RX_vect) {
+    PORTD ^= 1 << 3;
+    uint8_t read_char = UDR0,  message;
+    if(read_char == my_address) {
+        if((address_of_message % 2) == 0) {
+            now = vision_result[address_of_message / 2];
+            message = now & 0xFF;
+            now >>= 8;
+            now &= 0xFF;
+        } else {
+            message = now;
+        }
+        address_of_message = (address_of_message + 1) % 6;
+        sendc(message);
+    }
 }
 
 ISR(PCINT0_vect) {
-	pinstate = PINB;
-	changed_bits = pinstate ^ portb_history;
-	portb_history = pinstate;  
-	for (ct = 0; ct < 4; ct++) {
-		if (changed_bits & (1 << ct)) { 
-       		if (pinstate & (1 << ct)) { 
-				starts_of_pulses[ct] = TCNT1; 
-			} else {
-				ends_of_pulses[ct] = TCNT1;
-				lenghts_of_pulses[ct] = abs(ends_of_pulses[ct] - starts_of_pulses[ct]);
-			}
-		}
+    pinstate = PINB;
+    changed_bits = pinstate ^ portb_history;
+    portb_history = pinstate;
+    for (ct = 0; ct < 4; ct++) {
+        if (changed_bits & (1 << ct)) {
+            if (pinstate & (1 << ct)) {
+                starts_of_pulses[ct] = TCNT1;
+            } else {
+                ends_of_pulses[ct] = TCNT1;
+                if(ends_of_pulses[ct] < starts_of_pulses[ct]) {
+                    ends_of_pulses[ct] += TIMER_OVERFLOW;
+                }
+                lenghts_of_pulses[ct] = ends_of_pulses[ct]
+                                      -    starts_of_pulses[ct];
```

Comment:
Please remove the spaces between the minus sign (`-`) and `starts_of_pulses[ct];`.